File: 01-two-sum/typescript/two_sum.ts
// Two Sum Problem Solutions
// Brute Force and Optimal Approaches

function twoSumBruteForce(nums: number[], target: number): number[] {
    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                return [i, j];
            }
        }
    }
    return [];
}

function twoSumOptimal(nums: number[], target: number): number[] {
    const visited = new Map<number, number>();
    for (let i = 0; i < nums.length; i++) {
        const remaining = target - nums[i];
        if (visited.has(remaining)) {
            const index = visited.get(remaining);
            if (index !== undefined) {
                return [index, i]; // Return in consistent order
            }
        }
        visited.set(nums[i], i);
    }
    return [];
}

// Test the functions
const nums: number[] = [2, 7, 11, 15];
const target: number = 9;

console.log("Brute Force");
console.log(twoSumBruteForce(nums, target));
console.log("Optimal");
console.log(twoSumOptimal(nums, target));

File: 01-two-sum/go/two_sum.go
package main

import "fmt"

func main()  {
	nums := []int{2, 7, 11, 15}
	target := 9

	fmt.Println("Brute Force")
	fmt.Println(twoSumBruteForce(nums, target))

	fmt.Println("Optimal")
	fmt.Println(twoSumOptimal(nums, target))

}

func twoSumBruteForce(nums []int, target int) []int	{
	for i:=0; i<len(nums); i++ {
		for j:=i+1; j<len(nums); j++ {
			if nums[i]+nums[j] == target {
				return []int{i, j}
			}
		}
	}
	return []int{}
}

func twoSumOptimal(nums []int, target int) []int {
	visited := make(map[int]int)

	for i:=0; i<len(nums); i++ {
		remaining := target - nums[i]
		if j, exist := visited[remaining]; exist {
			return []int{i, j}
		}
		visited[nums[i]] = i;
	}
	return []int{}
}